app/services/proofing/base.rb

```diff
@@ -0,0 +1,99 @@
+require 'set'
+
+module Proofing
+  class Base
+    @vendor_name = nil
+    @required_attributes = []
+    @optional_attributes = []
+    @stage = nil
+
+    class << self
+      attr_reader :proofer
+
+      def vendor_name(name = nil)
+        @vendor_name = name || @vendor_name
+      end
+
+      def required_attributes(*required_attributes)
+        return @required_attributes || [] if required_attributes.empty?
+        @required_attributes = required_attributes
+      end
+
+      def optional_attributes(*optional_attributes)
+        return @optional_attributes || [] if optional_attributes.empty?
+        @optional_attributes = optional_attributes
+      end
+
+      def attributes
+        required_attributes.concat optional_attributes
+      end
+
+      def stage(stage = nil)
+        @stage = stage || @stage
+      end
+
+      def proof(sym = nil, &block)
+        @proofer = sym || block
+      end
+    end
+
+    def proof(applicant)
+      vendor_applicant = restrict_attributes(applicant)
+      validate_attributes(vendor_applicant)
+      result = Proofing::Result.new
+      execute_proof(proofer, vendor_applicant, result)
+      result
+    rescue StandardError => exception
```

Comment:
this is the same as:

```suggestion
    rescue => exception
```